# Проект по SQL

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


# Описание данных:

Таблица `books`:

    - book_id — идентификатор книги;
    - author_id — идентификатор автора;
    - title — название книги;
    - num_pages — количество страниц;
    - publication_date — дата публикации книги;
    - publication_date — идентификатор издателя.
    
Таблица `authors`:
    
    - author_id — идентификатор автора;
    - author — имя автора.
    
Таблица `publishers`:
  
    - publisher_id — идентификатор издательства;
    - publisher — название издательства;
    
Таблица `ratings`:
    
    - rating_id — идентификатор оценки;
    - book_id — идентификатор книги;
    - username — имя пользователя, оставившего оценку;
    - rating — оценка книги.
    
Таблица `reviews`:
    
    - review_id — идентификатор обзора;
    - book_id — идентификатор книги;
    - username — имя автора обзора;
    - text — текст обзора.
    
# Ход исследования:

1) Опишите цели исследования;

2) Исследуйте таблицы — выведите первые строки;

3) Сделайте по одному SQL-запросу для решения каждого задания;

4) Выведите результаты каждого запроса в тетрадке;

5) Опишите выводы по каждой из решённых задач.

In [14]:
import pandas as pd
from sqlalchemy import text, create_engine

In [15]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [16]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Исследование таблиц

In [17]:
query = '''
SELECT * FROM books
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [18]:
query = '''
SELECT * FROM authors
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [19]:
query = '''
SELECT * FROM publishers
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [20]:
query = '''
SELECT * FROM ratings
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [21]:
query = '''
SELECT * FROM reviews
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Комментарий**
Во всех таблицах есть данные и названия столбцов совпадают с заданными.

# Решение заданий с выводами

In [22]:
#Посчитайте, сколько книг вышло после 1 января 2000 года;
# Задача 1
query = '''
SELECT count(*) FROM books
where publication_date > '2000-01-01'
LIMIT 50
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**Вывод**

Ограничил дату, искал книги, вышедшие после 1 января 2000 года - таких книг 819.

In [23]:
#Для каждой книги посчитайте количество обзоров и среднюю оценку;
# Задача 2
query = '''
SELECT b.book_id, b.title, count(distinct review_id) as count_reviews, avg(rating) as avg_rating FROM books as b
left join reviews as r on b.book_id = r.book_id
left join ratings as rat on b.book_id = rat.book_id
group by b.book_id, b.title

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     count_reviews  avg_rating  
0                2    3.666667  
1                1    2.500000  
2                3    4.666667  
3                2    4.500000  
4                4    4.000000  
..             ...         ...  
995              3    3.666667  
996              3    3.400000  
997              4    3.200000  
998              2    4.500000  
999              4    3.833333  

[1000 rows x 4 columns]

**Вывод**

По условию задачи вывел название книги и его ид, количество обзоров на книгу и средняя оценка на книгу. Вывел путём присоединения таблиц книг, рейтинга и обзоров. Далее скриппировал таблицу по названию и ид книг и нашел количество отзывов и средний рейтинг книги.

In [24]:
#Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
# Задача 3
query = '''
SELECT publisher, p.publisher_id, count(*) as count_book FROM books as b
left join publishers as p on b.publisher_id = p.publisher_id
where num_pages > 50
group by p.publisher, p.publisher_id
order by count_book desc
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  publisher_id  count_book
0  Penguin Books           212          42

**Вывод**

Определил издательство, которое выпустило больше всего книг толще 50 страниц.

Выгрузил данные следующим способом - объединил таблицы книг и издательств, ограничил данные количеством страниц больше 50, сгрупиировал данные по издателю и его ид и стал искать издателей по убыванию количества изданных книг. Вывел 1 значение, чтобы указать издательство, которое выпустило наибольшее число книг.

In [25]:
#Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
# Задача 4
query = '''
with t as (

SELECT a.author, count(distinct r.rating_id) as count_rating, avg(r.rating) as avg_rating FROM books as b
left join authors as a on b.author_id = a.author_id
left join ratings as r on b.book_id = r.book_id 
group by a.author
having count(distinct r.rating_id) > 50
order by avg_rating desc
LIMIT 10
)

select * from t
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  count_rating  avg_rating
0         J.K. Rowling/Mary GrandPré           312    4.288462
1                    Agatha Christie            53    4.283019
2  Markus Zusak/Cao Xuân Việt Khương            53    4.264151
3                     J.R.R. Tolkien           166    4.240964
4           Roald Dahl/Quentin Blake            62    4.209677
5                  Louisa May Alcott            54    4.203704
6                       Rick Riordan            84    4.130952
7                      Arthur Golden            56    4.107143
8                       Stephen King           106    4.009434
9                       John Grisham            70    3.971429

In [28]:
#Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
# Задача 4
query = '''
--with t as (

SELECT a.author, r.rating_id, count(distinct r.rating_id) as count_rating, avg(r.rating) as avg_rating FROM books as b
left join authors as a on b.author_id = a.author_id
left join ratings as r on b.book_id = r.book_id 
group by a.author
having count(distinct r.rating_id) > 50
order by avg_rating desc
LIMIT 10
--)

--select * from t
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

Empty DataFrame
Columns: [author, rating_id, count_rating, avg_rating]
Index: []

**Вывод**

В данной задаче я вывел два поля - автора и средняя оценка книг автора.

Для вывода этих данных было сделано следующее: обьединил три таблицы - книги, автор, рейтинг. Далее сгруппировал данные по автору, и ограничил количеством оценок больше 50. Далее вывел данные об средних оценках книг у автора по убыванию и вывел первую строчку, так как необходимо было указать автора с самой высокой средней оценкой книг.

In [13]:
#Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
# Задача 5
query = '''
with tt as (
select r.username, count(distinct r.review_id) as count_reviews FROM ratings as rat
join reviews as r on rat.username = r.username
where rat.username in (
    select rat.username from ratings as rat
    group by rat.username
    having count(rat.rating_id) > 48
    )
group by r.username)

select avg(count_reviews) from tt
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

**Вывод**

В данной задаче я посчитал среднее количество обзорово от пользователей, среди тех, кто поставил больше 48 оценок.

Для этой задачи задействовались только таблицы рейтинга и обзоров. Для начала я нашел количество уникальных ид обзоров, а так же сделал вложенную таблицу с именами пользователей, у которых более 48 оценок. Сгруппировал данные по имени пользователя и вывел среднее количество обзоров от пользователей.